src/neuroconv/tools/nwb_helpers/_models/_base_dataset_models.py

```diff
@@ -0,0 +1,115 @@
+"""Base Pydantic models for DatasetInfo and DatasetConfiguration."""
+from typing import Any, Dict, Tuple, Union
+
+import h5py
+import numcodecs
+import numpy as np
+from pydantic import BaseModel, Field, root_validator, validator
+
+
+class DatasetInfo(BaseModel):
+    # TODO: When using Pydantic v2, replace with
+    # model_config = ConfigDict(allow_mutation=False)
+    class Config:  # noqa: D106
+        allow_mutation = False
+        arbitrary_types_allowed = True
+
+    object_id: str = Field(description="The UUID of the neurodata object containing the dataset.")
+    location: str = Field(description="The relative location of the this dataset within the in-memory NWBFile.")
+    full_shape: Tuple[int, ...] = Field(description="The maximum shape of the entire dataset.")
+    dtype: np.dtype = Field(  # TODO: When using Pydantic v2, replace np.dtype with InstanceOf[np.dtype]
+        description="The data type of elements of this dataset."
+    )
+
+    def __hash__(self):
+        """To allow instances of this class to be used as keys in dictionaries."""
+        return hash((type(self),) + tuple(self.__dict__.values()))
+
+    def __str__(self) -> str:
+        """Not overriding __repr__ as this is intended to render only when wrapped in print()."""
```

Comment:
Can you elaborate on why we need to modify __str__ but you need to keep __repr__?

Like, obviously you want to keep __repr__ intact for some reason but the reason here is not clear.

Reply:
Added enhanced description

Reason being two-fold; a standard `repr` is intended to be slightly more machine readable / a more basic representation of the true object state. But then also because an iterable of these objects, such as a `List[DataSetInfo]`, would print out the nested representations, which only look good when using the basic `repr` (that is, this fancy string print-out does not look good when nested in another container).

Reply:
Thanks. Most of this was resolved in an in-person talk. So, this still does not play as intended in a notebook because notebooks use __repr__ when outputing a value instead of print.

I don't think that solving this to have good representation in all possible environments is a good use of our  time. So let's leave it as it is, right now you are the only playing with this and I think it makes sense that you chooose the representation that you find more useful. We can come back to this in the future if needed.